packages/mwp-core/src/renderers/server-render.jsx

```diff
@@ -203,125 +203,133 @@ const makeRenderer = (
+					entryPath: url.pathname, // the path that the user entered the app on
+					media: getMedia(userAgent, userAgentDevice),
+				},
+			};
 
-	// otherwise render using the API and React router
-	const addFlags = store => {
-		const memberObj = (store.getState().api.self || {}).value || {};
-		return request.server.plugins['mwp-app-route']
-			.getFlags(memberObj)
-			.then(flags =>
-				store.dispatch({
-					type: 'UPDATE_FLAGS',
-					payload: flags,
-				})
+			const createStore = getServerCreateStore(
+				getFindMatches(resolvedRoutes, basename),
+				middleware,
+				request
 			);
-	};
-	const checkReady = state =>
-		state.preRenderChecklist.every(isReady => isReady);
-	const populateStore = store =>
-		new Promise((resolve, reject) => {
-			// dispatch SERVER_RENDER to kick off API middleware
-			store.dispatch({ type: SERVER_RENDER, payload: url });
+			return Promise.resolve(createStore(reducer, initialState));
+		};
+
+		// otherwise render using the API and React router
+		const addFlags = store => {
+			const memberObj = (store.getState().api.self || {}).value || {};
+			return request.server.plugins['mwp-app-route']
+				.getFlags(memberObj)
+				.then(flags =>
+					store.dispatch({
+						type: 'UPDATE_FLAGS',
+						payload: flags,
+					})
+				);
+		};
+		const checkReady = state =>
+			state.preRenderChecklist.every(isReady => isReady);
+		const populateStore = store =>
+			new Promise((resolve, reject) => {
+				// dispatch SERVER_RENDER to kick off API middleware
+				store.dispatch({ type: SERVER_RENDER, payload: url });
 
-			if (checkReady(store.getState())) {
-				addFlags(store).then(() => {
-					resolve(store);
-				});
-				return;
-			}
-			const unsubscribe = store.subscribe(() => {
 				if (checkReady(store.getState())) {
 					addFlags(store).then(() => {
 						resolve(store);
-						unsubscribe();
 					});
+					return;
 				}
+				const unsubscribe = store.subscribe(() => {
+					if (checkReady(store.getState())) {
+						addFlags(store).then(() => {
+							resolve(store);
+							unsubscribe();
+						});
+					}
+				});
 			});
-		});
 
-	return initializeStore().then(store => {
-		if ('skeleton' in request.query) {
-			// render skeleton if requested - the store is ready
-			return {
-				result: getHtml(
-					<Dom
-						basename={basename}
-						head={Helmet.rewind()}
-						initialState={store.getState()}
-						scripts={scripts}
-						cssLinks={cssLinks}
-					/>
-				),
-				statusCode: 200,
-			};
-		}
-		return populateStore(store).then(
-			store =>
-				// create tracer and immediately invoke the resulting function.
-				// trace should start before rendering, finish after rendering
-				newrelic.createTracer('serverRender', getRouterRenderer)({
-					routes,
-					store,
-					location: url,
-					basename,
-					scripts,
-					cssLinks,
-					userAgent,
-				}) // immediately invoke callback
-		);
-	});
+		return routesPromise.then(resolvedRoutes => {
```

Comment:
this promise is 'resolved' on server startup, but we can attach a `.then` handler for each new request and it will get the resolved value 'instantly'